File: api/Trackster.Api/Features/Users/UserMapper.cs
using Trackster.Api.Data.Records;
using Trackster.Api.Features.Users.Types;

namespace Trackster.Api.Features.Users;

public class UserMapper
{
    public static UserRecord MapRecord(User user)
    {
        return new UserRecord
        {
            Identifier = user.Identifier,
            Email = user.Email,
            Username = user.Username,
            Password = user.Password,
            CreatedAt = user.CreatedAt,
            UpdatedAt = user.UpdatedAt,
            ThirdPartyIntegrations = user.ThirdPartyIntegrations.Select(ThirdPartyIntegrationMapper.MapRecord).ToList()
        };
    }

    public static User Map(UserRecord record)
    {
        return new User
        {
            Identifier = record.Identifier,
            Email = record.Email,
            Username = record.Username,
            Password = record.Password,
            CreatedAt = record.CreatedAt,
            UpdatedAt = record.UpdatedAt,
            ThirdPartyIntegrations = record.ThirdPartyIntegrations.Select(ThirdPartyIntegrationMapper.Map).ToList()
        };
    }
}

File: src/services/user-service/user.mapper.ts
import { IUser } from './types/user.type';

export class UserMapper {
    public static map(user: any): IUser {
        return {
            identifier: user.Identifier,
            username: user.Username,
            email: user.Email,
            createdAt: user.CreatedAt,
            thirdPartyIntegrations: user.ThirdPartyIntegrations.map((x: any) => {
                return {
                    identifier: x.Identifier,
                    provider: x.Provider,
                };
            }),
        };
    }

    public static mapRecord(user: IUser): any {
        return {
            Identifier: user.identifier,
            Username: user.username,
            Email: user.email,
            CreatedAt: user.createdAt,
            ThirdPartyIntegrations: user.thirdPartyIntegrations.map((x) => {
                return {
                    Identifier: x.identifier,
                    Provider: x.provider,
                };
            }),
        };
    }
}